contracts/ModuleRegistry.sol

```diff
@@ -118,29 +120,36 @@ contract ModuleRegistry is IModuleRegistry, EternalStorage {
      * @dev The feature switch for custom modules is labelled "customModulesAllowed"
      * @param _moduleFactory is the address of the relevant module factory
      */
-    function useModule(address _moduleFactory) external {
-        // This if statement is required to be able to add modules from the token proxy contract during deployment
-        if (ISecurityTokenRegistry(getAddressValue(Encoder.getKey("securityTokenRegistry"))).isSecurityToken(msg.sender)) {
-            if (IFeatureRegistry(getAddressValue(Encoder.getKey("featureRegistry"))).getFeatureStatus("customModulesAllowed")) {
-                require(
-                    getBoolValue(Encoder.getKey("verified", _moduleFactory)) || IOwnable(_moduleFactory).owner() == IOwnable(msg.sender).owner(),
-                    "ModuleFactory must be verified or SecurityToken owner must be ModuleFactory owner"
-                );
-            } else {
-                require(getBoolValue(Encoder.getKey("verified", _moduleFactory)), "ModuleFactory must be verified");
-            }
-            require(_isCompatibleModule(_moduleFactory, msg.sender), "Version should within the compatible range of ST");
+    function useModule(address _moduleFactory, bool _isUpgrade) external {
+        if (IFeatureRegistry(getAddressValue(Encoder.getKey("featureRegistry"))).getFeatureStatus("customModulesAllowed")) {
+            require(
+                getBoolValue(Encoder.getKey("verified", _moduleFactory)) || IOwnable(_moduleFactory).owner() == IOwnable(msg.sender).owner(),
+                "ModuleFactory must be verified or SecurityToken owner must be ModuleFactory owner"
+            );
+        } else {
+            require(getBoolValue(Encoder.getKey("verified", _moduleFactory)), "ModuleFactory must be verified");
+        }
+        require(isCompatibleModule(_moduleFactory, msg.sender), "Incompatible versions");
```

Comment:
Yep - I did think about this and the MR has no particular permissions I believe, but for safety sake (and possibly future bugs) have moved it into the if clause.

Reply:
Actually - unresolving this one - moving this check inside the if clause means that module bounds are not checked when we add the GTM from the STF, as the token has not yet been registered in the STR. I can't see anything malicious that could happen from forcing the MTR to call out to an arbitrary contract but equally a STF should never be setup in such a way as to try and include an incompatible module. Neither ideal - maybe this area can be re-worked a bit at a higher level to improve the logic.

Reply:
I put this logic back inside the `if` clause. I also perform the check (compatible modules) if the caller is the current STFactory. This still leaves a gap if the caller is an old STFactory (i.e. someone has used generateST passing in an old protocol version). This could be solved if we add something like: isSTFactory(address) into the STR, but would need the STR to maintain a list of active STFactories. Will add an Asana for this but not covering in this PR.

Reply:
>I also perform the check (compatible modules) if the caller is the current STFactory

Why do we need this check? When STF calls addModule on the security token to add the GTM, the security token will call Module Factory to verify bounds so the msg.sender that MR gets will be of the new (unregistered) ST instead of STF. Marking this as unresolved as I couldn't understand the reasoning.

Reply:
You are correct - the code makes no difference as msg.sender is the ST, not the STF - my mistake - will remove this line again, and we are back in the (current) situation that STF module deployments don't check bounds. We just need to be careful about making sure that STF's are "self-consistent".